### PR TITLE
add in predictive back snippets for NavHost and predictive back handler

### PR DESCRIPTION
**Pull request by trambui09**
Add in the predictive back snippets for NavHost implementation and the Predictive Back Handler for a navigation drawer

### CODE REVIEW

**Review by secondsun — APPROVED** [review]

**Review by ashnohe — APPROVED** [review]

**Comment by ashnohe** [comment]
@googlebot hello

File: compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+@Composable
+private fun PredictiveBackOverrideExit(
+    modifier: Modifier,
+) {
+    val navController = rememberNavController()
+
+    // [START android_compose_predictiveback_navhost]
+    NavHost(
+        navController = navController,
+        startDestination = "home",
```
Comment (dturner) [inline compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt:59]: This should use type safe routes. To do this declare: 

```
@Serializable data object Home
@Serializable data object Settings
```

before the function. 

Then to use it do `startDestination = Home`

File: compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            )
+        },
+        popEnterTransition = {
+            EnterTransition.None
+        },
+        modifier = modifier,
+    )
+    // [END android_compose_predictiveback_navhost]
+    {
+        composable("home") {
```
Comment (dturner) [inline compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt:73]: `composable<Home>`

File: compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            HomeScreen(
+                modifier = modifier,
+                navController = navController,
+            )
+        }
+        composable("settings") {
```
Comment (dturner) [inline compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt:79]: `composable<Settings>`

File: compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+    // [START android_compose_predictivebackhandler_basic]
+    PredictiveBackHandler(true) { progress: Flow<BackEventCompat> ->
+        // code for gesture back started
+        try {
+            progress.collect { backEvent ->
+                // code for progress
+                boxScale = 1F - (1F * backEvent.progress)
+            }
+            // code for completion
```
Comment (dturner) [inline compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt:121]: Shouldn't this have a completion state? From the looks of it, it should be `boxScale = 0F`

File: compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+import kotlinx.coroutines.flow.Flow
+
+@Composable
+private fun PredictiveBackOverrideExit(
+    modifier: Modifier,
+) {
+    val navController = rememberNavController()
+
+    // [START android_compose_predictiveback_navhost]
+    NavHost(
```
Comment (dturner) [inline compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt:57]: Not sure of the context in which this is being used but I think it's important to mention that `NavHost` supports predictive back out of the box. There's nothing extra required, aside from defining the animations you want to use between destinations.

**Review by dturner — COMMENTED** [review]
Sorry I wasn't able to review before merging. Added a few comments and suggestions.

File: compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt (hunk 1 above)
Comment (trambui09) [reply compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt:59]: Ahhh, thanks for the tip of changing them to safe routes, Don!

File: compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt (hunk 4 above)
Comment (trambui09) [reply compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt:121]: Yep, it should have a completion state. Added it into the https://github.com/android/snippets/pull/399 PR.

File: compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt (hunk 5 above)
Comment (trambui09) [reply compose/snippets/src/main/java/com/example/compose/snippets/predictiveback/PredictiveBackSnippets.kt:57]: Yes, we do mention NavHost supports PB out of the box. This snippets shows a custom in-app animations example for PB

**Pull request merged**